src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,207 @@
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(peerId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established as needed.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::RemoveUnicastBinding(FabricIndex fabric, NodeId node)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+    PeerId peer                      = fabricInfo->GetPeerIdForNode(node);
+    PendingNotificationEntry * entry = mPendingNotificationMap.FindEntry(peer);
+
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(peer);
+    mPendingNotificationMap.RemoveEntry(entry);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingManager::NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster, void * context)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        EmberBindingTableEntry entry;
+
+        if (emberGetBinding(i, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.local == endpoint &&
+            entry.clusterId == cluster)
+        {
+            if (entry.type == EMBER_UNICAST_BINDING)
+            {
+                FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(entry.fabricIndex);
+                VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+                PeerId peer                         = fabricInfo->GetPeerIdForNode(entry.nodeId);
+                OperationalDeviceProxy * peerDevice = mAppServer->GetCASESessionManager()->FindExistingSession(peer);
+                if (peerDevice != nullptr && mBoundDeviceChangedHandler)
+                {
+                    // We already have an active connection
+                    mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, peerDevice, NullOptional, context);
+                }
+                else
+                {
+                    // Enqueue pending cluster and establish connection
+                    ReturnErrorOnFailure(
+                        EnqueueUnicastNotification(entry.fabricIndex, entry.nodeId, entry.local, entry.clusterId, context));
+                    ReturnErrorOnFailure(CreateUnicastBinding(entry.fabricIndex, entry.nodeId));
+                }
+            }
+            // TODO: Add group binding support
```

Comment:
Why can't this just be:
```suggestion
            else if (entry.type == EMBER_MULTICAST_BINDING) 
            {
                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, nullptr, MakeOptional<entry.groupId>, context);
            }
```
?  It's not great, in that entry.remote is nonsense in the group case, so the `mBoundDeviceChangedHandler` signature is still pretty broken in spite of @mrjerryjohns and me asking to make it work for groups too, but it'll be ok-ish....